File: Serene/Serene.Script/Northwind/OrderDetail/OrderDetailDialog.cs

namespace Serene.Northwind
{
    using Common;
    using Serenity;

    [FormKey("Northwind.OrderDetail"), LocalTextPrefix("Northwind.OrderDetail")]
    public class OrderDetailDialog : GridEditorDialog<OrderDetailRow>
    {
        protected OrderDetailForm form;

        public OrderDetailDialog()
        {
            form = new OrderDetailForm(this.IdPrefix);

            form.ProductID.ChangeSelect2(e => {
                var productID = form.ProductID.Value.ToInt32();
                if (productID != null)
                    form.UnitPrice.Value = (double?)ProductRow.Lookup.ItemById[productID.Value].UnitPrice;
            });

            form.Discount.AddValidationRule(this.uniqueName, e => {
                if (form.UnitPrice.Value != null &&
                    form.Quantity.Value != null &&
                    form.Discount.Value != null &&
                    form.Discount.Value.Value > 0 &&
                    form.Discount.Value.Value > form.UnitPrice.Value.Value * form.Quantity.Value.Value)
                {
                    return "Discount can't be higher than total price!";
                }

                return null;
            });
        }
    }
}

File: Serene/Serene.Web/Modules/Northwind/OrderDetail/OrderDetailDialog.ts
/// <reference path="../../Common/Helpers/GridEditorDialog.ts" />

namespace Serene.Northwind {

    @Serenity.Decorators.registerClass()
    export class OrderDetailDialog extends Common.GridEditorDialog<OrderDetailRow> {
        protected getFormKey() { return OrderDetailForm.formKey; }
        protected getLocalTextPrefix() { return OrderDetailRow.localTextPrefix; }

        protected form: OrderDetailForm;

        constructor() {
            super();

            this.form = new OrderDetailForm(this.idPrefix);

            this.form.ProductID.changeSelect2(e => {
                var productID = Q.toId(this.form.ProductID.value);
                if (productID != null) {
                    this.form.UnitPrice.value = ProductRow.getLookup().itemById[productID].UnitPrice;
                }
            });

            this.form.Discount.addValidationRule(this.uniqueName, e => {
                var price = this.form.UnitPrice.value;
                var quantity = this.form.Quantity.value;
                var discount = this.form.Discount.value;
                if (price != null && quantity != null && discount != null &&
                    discount > 0 && discount >= price * quantity) {
                    return "Discount can't be higher than total price!";
                }
            });
        }
    }
}